### PR TITLE
S2439

### PR DESCRIPTION
**Pull request by yeunga**
Added CadcAlmaResolver to support ALMA.

### CODE REVIEW

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        sanitize(fileID);
+
+        // trim the archive to 6 characters
+        // TODO: Remove this trim when AD supports longer archive names
+        if (archive.length() > 6) {
+            archive = archive.substring(0, 6);
+        }
+        
+        return archive + "/" + fileID;
+    }
```
Comment (brianmajor) [inline caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java]: Perhaps you should make this method package protected so you can write a unit test against it.

Comment (yeunga) [reply caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java]: There is already a unit test for CadcAlmaResolver. I can add more asserts, e.g. check that there is no 'namespace' in the returned URL.

**Comment by brianmajor** [comment]
On Fri, Mar 29, 2019 at 9:46 AM yeunga <notifications@github.com> wrote:

> *@yeunga* commented on this pull request.
> ------------------------------
>
> In
> caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java
> <https://github.com/opencadc/caom2/pull/101#discussion_r270492031>:
>
> > +
> +        if (!StringUtil.hasText(fileID)) {
> +            throw new IllegalArgumentException("cannot extract fileID from " + uri);
> +        }
> +
> +        sanitize(archive);
> +        sanitize(fileID);
> +
> +        // trim the archive to 6 characters
> +        // TODO: Remove this trim when AD supports longer archive names
> +        if (archive.length() > 6) {
> +            archive = archive.substring(0, 6);
> +        }
> +
> +        return archive + "/" + fileID;
> +    }
>
> There is already a unit test for CadcAlmaResolver. I can add more asserts,
> e.g. check that there is no 'namespace' in the returned URL.
>
That won't work because the URL you'll receive may or may not have a
namespace, depending on what the ad resolver does.  Currently, the URL you
get does have a 'namespace'.  I think localizing the test to that method
would be better and less fragile.

Brian

File: caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java (hunk 1 above)
Comment (yeunga) [reply caom2-artifact-resolvers/src/main/java/ca/nrc/cadc/caom2/artifact/resolvers/CadcAlmaResolver.java]: In the unit test, if we set the artifact URI to "alma:ALMA/some/name/space/bar.fits", we can expect the returned URL to end with ALMA/bar.fits. This verifies that 'some/name/space/' has been removed.

**Comment by brianmajor** [comment]
Okay.

**Pull request merged**